File: leetsync/2695-find-score-of-an-array-after-marking-all-elements/find-score-of-an-array-after-marking-all-elements.java
class Solution {
    public long findScore(int[] nums) {
        int n = nums.length;
        boolean[] visited = new boolean[n];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[0] != b[0]) {
                return Integer.compare(a[0], b[0]);
            } else {
                return Integer.compare(a[1], b[1]);
            }
        });
        for (int i = 0; i < n; i++) {
            pq.add(new int[] { nums[i], i });
        }

        long score = 0;
        while (!pq.isEmpty()) {
            int[] element = pq.poll();
            int value = element[0];
            int index = element[1];

            if (!visited[index]) {
                visited[index] = true;
                score += value;
                // Mark adjacent indices as visited
                if (index - 1 >= 0) {
                    visited[index - 1] = true;
                }
                if (index + 1 < n) {
                    visited[index + 1] = true;
                }
            }
        }
        return score;
    }
}

File: leetsync/2695-find-score-of-an-array-after-marking-all-elements/find-score-of-an-array-after-marking-all-elements.go
func findScore(nums []int) int64 {
	elements := make([]element, len(nums))
	for i := 0; i < len(nums); i++ {
		elements[i] = element{
			value: nums[i],
			index: i,
		}
	}
	// use a map to check whether an element has been marked
	marked := make(map[int]bool)
	h := minHeap(elements)
	heap.Init(&h)

	var score int
	for h.Len() > 0 {
		item := heap.Pop(&h).(element)
		if _, ok := marked[item.index]; ok {
			continue
		}
		score += item.value
		marked[item.index] = true
		marked[item.index+1] = true
		marked[item.index-1] = true
	}
	return int64(score)
}

type element struct {
	value int
	index int
}

type minHeap []element

func (h minHeap) Len() int {
	return len(h)
}

func (h minHeap) Less(i, j int) bool {
	if h[i].value == h[j].value {
		return h[i].index < h[j].index
	}
	return h[i].value < h[j].value
}

func (h minHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *minHeap) Push(x any) {
	*h = append(*h, x.(element))
}

func (h *minHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}